glib/src/unichar.rs

```diff
@@ -0,0 +1,237 @@
+
+    #[doc(alias = "g_unichar_decompose")]
+    #[doc(alias = "unichar_decompose")]
+    fn decompose(self) -> DecompositionResult;
+
+    #[doc(alias = "g_unichar_compose")]
+    #[doc(alias = "unichar_compose")]
+    fn compose(a: char, b: char) -> Option<char>;
+}
+
+impl Unichar for char {
+    fn unicode_type(self) -> UnicodeType {
+        unsafe { UnicodeType::from_glib(ffi::g_unichar_type(self as _)) }
+    }
+
+    fn break_type(self) -> UnicodeBreakType {
+        unsafe { UnicodeBreakType::from_glib(ffi::g_unichar_break_type(self as _)) }
+    }
+
+    fn script(self) -> UnicodeScript {
+        unsafe { UnicodeScript::from_glib(ffi::g_unichar_get_script(self as _)) }
+    }
+
+    fn combining_class(self) -> i32 {
+        unsafe { ffi::g_unichar_combining_class(self as _) }
+    }
+
+    fn is_mark(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_ismark(self as _)) }
+    }
+
+    fn is_graphical(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_isgraph(self as _)) }
+    }
+
+    fn is_punctuation(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_ispunct(self as _)) }
+    }
+
+    fn is_titlecase(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_istitle(self as _)) }
+    }
+
+    fn is_defined(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_isdefined(self as _)) }
+    }
+
+    fn is_wide(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_iswide(self as _)) }
+    }
+
+    fn is_wide_cjk(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_iswide_cjk(self as _)) }
+    }
+
+    fn is_zero_width(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_iszerowidth(self as _)) }
+    }
+
+    fn to_titlecase(self) -> char {
+        from_unichar(unsafe { ffi::g_unichar_totitle(self as _) })
+    }
+
+    fn mirror_char(self) -> Option<char> {
+        let mut mirrored = 0u32;
+        let res =
+            unsafe { bool::from_glib(ffi::g_unichar_get_mirror_char(self as _, &mut mirrored)) };
+        if res {
+            Some(from_unichar(mirrored))
+        } else {
+            None
+        }
+    }
+
+    fn fully_decompose(self, decomposition_kind: DecompositionKind) -> Vec<char> {
+        let buffer_len = ffi::G_UNICHAR_MAX_DECOMPOSITION_LENGTH as usize;
+        let mut buffer = vec![char::from(0); buffer_len];
+        let decomposition_length = unsafe {
+            ffi::g_unichar_fully_decompose(
+                self as _,
+                // NOTE: This assumes that glib will only ever write valid unicode codepoints to the
+                //       buffer.
+                decomposition_kind.into_bool().into_glib(),
+                buffer.as_mut_ptr().cast(),
+                buffer_len,
+            )
+        };
+        debug_assert!(decomposition_length <= buffer_len);
+        unsafe { buffer.set_len(decomposition_length) };
```

Comment:
This feels kinda sketch but I think it's fine?

Reply:
Yes but I would go one step further. Use `Vec::with_capacity()` above to create the `Vec` (and don't zero-initialize it).